### PR TITLE
MRPHS-4531: Network listing should not show "Uplink PortGroup" while associating project with vmware cloud

### PR DESCRIPTION
**Pull request by avnish30jn**
**Jira-id**

MRPHS-4531

**Problem**

Network listing should not show "Uplink PortGroup" while associating project with vmware cloud

**Solution**

Skipping port group if port group is uplink

**Unit Testing**

Done. Logs below:

screen shot 2018-02-02 at 5 32 50 pm

```
******************************************
"
time="02 Feb 18 11:58 UTC" level=info msg="[services.go:155::main.PrintAndValidateRequest]  C3ntry Server> Received request (get_network_list.reqId#62329):
 {Action:get_network_list Args:{"cloud_provider_id":18,"cloud_type":"vsphere","Insecure":true,"Username":"user@vsphere.local","Password":"passwd","Host":"starling-vcenter.gsintlab.com","datacenter":"Starling-DC","filter":{"clusters":["Cluster1"],"hosts":["67.220.186.130","67.220.186.131"]}} Id:18 ReplyTo: CorrelationId:}"
time="02 Feb 18 11:58 UTC" level=info msg="[services.go:157::main.PrintAndValidateRequest]
******************************************

"
time="02 Feb 18 11:58 UTC" level=debug msg="[c3modulemap.go:525::main.getNetworkList] Enter"
time="02 Feb 18 11:58 UTC" level=debug msg="[halo.go:1009::appos/c3ntry/halo.(*CloudArgs).GetNetworkList] Started Get network List"
time="02 Feb 18 11:58 UTC" level=debug msg="[halo.go:1020::appos/c3ntry/halo.(*CloudArgs).GetNetworkList] Call Halo->GetNetworkList()"
time="2018-02-02T11:58:30Z" level=debug msg="[actions.go:23::appos/vendor/github.com/apporbit/halo.glob..func1] Providing a vsphere cloud-object instance"
time="2018-02-02T11:58:30Z" level=debug msg="[vsphere.go:712::appos/vendor/github.com/apporbit/halo/cloud.(*VsphereCloud).GetNetworkList] [Enter]"
time="2018-02-02T11:58:30Z" level=debug msg="[vsphere.go:714::appos/vendor/github.com/apporbit/halo/cloud.(*VsphereCloud).GetNetworkList] [Creating vsphere.VM struct]"
time="2018-02-02T11:58:30Z" level=debug msg="[vsphere.go:322::appos/vendor/github.com/apporbit/halo/cloud.validateInput] Validate Inputs for action: GetNetworkList"
time="2018-02-02T11:58:30Z" level=debug msg="[vsphere.go:35::appos/vendor/github.com/apporbit/halo/cloud.mandatoryValidation] [Doing mandatory validation required for all the actions]"
time="2018-02-02T11:58:30Z" level=debug msg="[vsphere.go:52::appos/vendor/github.com/apporbit/halo/cloud.mandatoryValidation] [Exit: Mandatory Validations Succeeded]"
time="2018-02-02T11:58:30Z" level=debug msg="[vsphere.go:386::appos/vendor/github.com/apporbit/halo/cloud.validateInput] [Validation Succeeded]"
time="2018-02-02T11:58:30Z" level=debug msg="[vsphere.go:729::appos/vendor/github.com/apporbit/halo/cloud.(*VsphereCloud).GetNetworkList] [Calling Libretto->vsphere.GetDcNetworkList]"
time="2018-02-02T11:58:31Z" level=debug msg="[vsphere.go:742::appos/vendor/github.com/apporbit/halo/cloud.(*VsphereCloud).GetNetworkList] [Send networks as response object]"
time="02 Feb 18 11:58 UTC" level=info msg="[halo.go:1034::appos/c3ntry/halo.(*CloudArgs).GetNetworkList]  GetNetworkList succeeded"
time="02 Feb 18 11:58 UTC" level=debug msg="[c3modulemap.go:536::main.getNetworkList] Exit"
time="02 Feb 18 11:58 UTC" level=info msg="[services.go:162::main.ValidateAndPrintResponse]

=========================================
"
time="02 Feb 18 11:58 UTC" level=info msg="[services.go:163::main.ValidateAndPrintResponse]  C3ntry Server> Sending response (get_network_list.reqId#62329):
 {Result:[{"id":"network-1198","name":"DistributedPortGroup1"},{"id":"network-1533","name":"DummyPrivate-moved-to-distributed"},{"id":"network-15","name":"VM Network"},{"id":"network-15","name":"nativevmportgroup"},{"id":"network-15","name":"DistributedPrivateNetwork"}] Errcode:0 Id:18 ReplyTo: CorrelationId: Progress:100}"
time="02 Feb 18 11:58 UTC" level=info msg="[services.go:165::main.ValidateAndPrintResponse]
=========================================
```

### CODE REVIEW

File: virtualmachine/vsphere/vm.go [hunk 1]
```diff
@@ -1284,50 +1284,35 @@ func GetNetworkInHost(vm *VM) ([]map[string]string, error) {
...
+		networkMap map[string]string
+		network    mo.Network
+		portGroup  mo.DistributedVirtualPortgroup
 	)
 	networkList := make([]map[string]string, 0)
 
 	// Add all the networks in host to network list
 	for _, networkMor := range networkMo {
 		switch networkMor.Type {
 		case "Network":
```
Comment (nirav1007) [inline virtualmachine/vsphere/vm.go]: Do we need a default case in switch?

Comment (nirav1007) [inline virtualmachine/vsphere/vm.go:1296]: Do we need default case in switch for this?

File: virtualmachine/vsphere/vm.go [hunk 2]
```diff
@@ -1284,50 +1284,35 @@ func GetNetworkInHost(vm *VM) ([]map[string]string, error) {
...
-			if err != nil {
-				return nil, err
-			}
-			networkMap = map[string]string{"name": dvSwitch.Name, "id": network.Self.Value}
-		case "VmwareDistributedVirtualSwitch":
-			err := vm.collector.RetrieveOne(vm.ctx, networkMor, []string{"name"}, &vmwareDvSwitch)
-			if err != nil {
-				return nil, err
+			if portGroup.Config.Uplink == nil ||
+				*portGroup.Config.Uplink {
```
Comment (nirav1007) [inline virtualmachine/vsphere/vm.go:1311]: what the second part of if conditions checks?

**Review by nirav1007 — COMMENTED** [review]
Some comments inline

File: virtualmachine/vsphere/vm.go (hunk 2 above)
Comment (avnish30jn) [reply virtualmachine/vsphere/vm.go:1311]: the portgroups that are uplinks should not be listed. The 2nd if condition is to check if port group is uplink or not.

**Review by nirav1007 — APPROVED** [review]
LGTM

File: virtualmachine/vsphere/vm.go (hunk 1 above)
Comment (avnish30jn) [reply virtualmachine/vsphere/vm.go:1296]: in default case, there are other network types which are not supported, we need to skip those networks in the list.

**Pull request merged**